File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
import sys
from collections import deque

input = sys.stdin.readline

n, m = map(int, input().rstrip().split())

board = [0] * 101
visited = [False] * 101

for _ in range(n + m):
    a, b = map(int, input().rstrip().split())
    board[a] = b

queue = deque()

queue.append([1, 0])
visited[1] = True

while queue:
    cur = queue.popleft()
    pos = cur[0]
    cnt = cur[1]

    if pos == 100:
        print(cnt)
        quit()

    for i in range(7):
        nextPos = pos + i

        if nextPos > 100:
            continue

        if board[nextPos] != 0:
            nextPos = board[nextPos]

        if not visited[nextPos]:
            visited[nextPos] = True
            queue.append([nextPos, cnt + 1])

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());

		int[] board = new int[101];
		boolean[] visited = new boolean[101];

		for (int i = 0; i < n + m; i++) {
			st = new StringTokenizer(br.readLine());

			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());

			board[u] = v;
		}

		ArrayDeque<int[]> queue = new ArrayDeque<>();

		queue.add(new int[]{1, 0});
		visited[1] = true;

		while (!queue.isEmpty()) {
			int[] cur = queue.poll();
			int pos = cur[0];
			int count = cur[1];

			if (pos == 100) {
				System.out.println(count);
				return;
			}

			for (int i = 0; i < 7; i++) {
				int nextPos = pos + i;

				if (nextPos > 100) {
					continue;
				}

				if (board[nextPos] != 0) {
					nextPos = board[nextPos];
				}

				if (!visited[nextPos]) {
					visited[nextPos] = true;
					queue.add(new int[]{nextPos, count + 1});
				}
			}
		}
	}
}
